pkg/kine/drivers/generic/generic.go

```diff
@@ -381,6 +393,119 @@ func (d *Generic) queryRowPrepared(ctx context.Context, txName, sql string, prep
+
+		go d.watchBatchQueue(ctx)
+		go d.batchQueueWatcher(ctx)
+	}
+}
+
+func (d *Generic) batchQueueWatcher(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case item := <-d.addToBatchQueueCh:
+			d.batchingQueue = append(d.batchingQueue, item)
+			if len(d.batchingQueue) == d.BatchingMaxQueries {
+				d.flushCh <- struct{}{}
+			}
+		case count := <-d.removeFromBatchQueueCh:
+			d.batchingQueue = d.batchingQueue[count:]
+		}
+	}
+}
+
+func (d *Generic) processBatchedInserts(ctx context.Context, queue []BatchedInsert) ([]int64, error) {
+	valueStrings := make([]string, 0, len(queue))
+	// len(queue) * argument_count per insert
+	valueArgs := make([]interface{}, 0, len(queue)*8)
+
+	for _, insertItem := range queue {
+		valueStrings = append(valueStrings, "(?, ?, ?, ?, ?, ?, ?, ?)")
+		valueArgs = append(valueArgs, insertItem.args...)
+	}
+
+	// Maybe move this into the Generic struct like other queries
+	// Although since we are dynamically generating this we'd only store the template string there
+	stmt := fmt.Sprintf("INSERT INTO kine(name, created, deleted, create_revision, prev_revision, lease, value, old_value) VALUES%s", strings.Join(valueStrings, ","))
+
+	prepared, err := d.DB.Prepare(stmt)
+	if err != nil {
+		logrus.WithError(err).Error("Process batch queue error")
+		return nil, err
+	}
+
+	row, err := d.executePrepared(ctx, "batch_insert_last_insert_id_sql", stmt, prepared, valueArgs...)
+
+	if err != nil {
+		logrus.WithError(err).Error("Process batch queue error")
+		return nil, err
+	}
+
+	lastId, err := row.LastInsertId()
+
+	if err != nil {
+		logrus.WithError(err).Error("Process batch queue error")
+		return nil, err
+	}
+
+	var ids []int64
+
+	// Ideally instead of counting back from LastInsertId, sqlite supports RETURNING
+	// so something like "INSERT INTO .... RETURNING" id would be better.
+	// However db.Prepare complains about wrong syntax when RETURNING is used
+	for i := len(queue); i > 0; i-- {
+		ids = append(ids, lastId-int64(i)+1)
+	}
+
+	return ids, nil
+}
+
+func (d *Generic) processBatchQueue(ctx context.Context) {
+	if len(d.batchingQueue) > 0 {
+		dcq := make([]BatchedInsert, len(d.batchingQueue))
+		copy(dcq, d.batchingQueue)
+
+		ids, err := d.processBatchedInserts(ctx, dcq)
+		if err != nil {
+			logrus.WithError(err).Error("Process batch queue error")
+			return
+		}
+
+		for i, insertItem := range dcq {
+			insertItem.retCh <- ids[i]
+		}
+
+		d.removeFromBatchQueueCh <- len(dcq)
+	}
+}
+
+func (d *Generic) watchBatchQueue(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			d.processBatchQueue(ctx)
+			return
+		case <-d.flushCh:
+			d.processBatchQueue(ctx)
+		case <-time.After(d.BatchingInterval):
+			d.processBatchQueue(ctx)
+		}
+	}
+}
+
+func (d *Generic) batchQueryRowPrepared(ctx context.Context, txName, sqli string, prepared *sql.Stmt, args ...interface{}) (id int64) {
+	ret := make(chan int64)
```

Comment:
So that we don't block when writing to the channel

```suggestion
	ret := make(chan int64, 1)
```